File: 알고리즘 기본 구현/Tarjan's Algorithm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;

namespace Tarjan {
    using SCC = List<Vertex>;

    class Vertex : IComparable<Vertex> {
        public int Number { get; init; }
        public int? VisitedID { get; set; }
        public bool Finished { get; set; }
        public List<Vertex> Neighbourhood { get; private set; }

        public Vertex(int num) {
            Number = num;
            Neighbourhood = new List<Vertex>();
        }

        public int CompareTo(Vertex v) {
            return Number.CompareTo(v.Number);
        }
    }

    class Graph {
        private Vertex[] vertices;

        public Graph(int vertexCount) {
            vertices = new Vertex[vertexCount];
            for (int i = 0; i < vertexCount; ++i) {
                vertices[i] = new Vertex(i);
            }
        }

        public void AddEdge(int s, int e) {
            vertices[s].Neighbourhood.Add(vertices[e]);
        }

        public List<SCC> FindSCCs() {
            foreach (Vertex v in vertices) {
                v.VisitedID = null;
                v.Finished = false;
            }

            List<SCC> sccList = new List<SCC>();
            Stack<Vertex> recStack = new Stack<Vertex>();
            int id = 0;

            foreach (Vertex v in vertices) {
                if (v.VisitedID == null) {
                    FindParentByDFS(v, ref id, sccList, recStack);
                }
            }

            return sccList;
        }

        private Vertex FindParentByDFS(Vertex start, ref int id, List<SCC> sccList, Stack<Vertex> recStack) {
            Vertex parent = start;
            start.VisitedID = id++;
            recStack.Push(start);

            foreach (Vertex next in start.Neighbourhood) {
                if (next.VisitedID == null) {
                    Vertex found = FindParentByDFS(next, ref id, sccList, recStack);
                    parent = (found.VisitedID < parent.VisitedID ? found : parent);
                } else if (!next.Finished) {
                    parent = (next.VisitedID < parent.VisitedID ? next : parent);
                }
            }

            if (parent == start) {
                SCC scc = new SCC();
                Vertex top;
                do {
                    top = recStack.Pop();
                    top.Finished = true;
                    scc.Add(top);
                } while (top != parent);
                sccList.Add(scc);
            }

            return parent;
        }
    }
}

File: 알고리즘 기본 구현/CPP/Tarjan's Algorithm.cpp
#include <iostream>
#include <vector>
#include <functional>

int main() {
    std::cin.tie(nullptr)->sync_with_stdio(false);

    int n, e;
    std::cin >> n >> e;

    std::vector<std::vector<int>> graph(n+1);
    for (int i = 0; i < e; ++i) {
        int u, v;
        std::cin >> u >> v;
        graph[u].push_back(v);
    }

    using SCC = std::vector<int>;
    std::vector<SCC> sccs;
    std::vector<int> recStack, visited(n+1, 0);
    std::vector<bool> finished(n+1, false);
    int id = 1;

    std::function<int(int)> findParentByDFS = [&](int start) -> int {
        int parent = start;
        visited[start] = id++;
        recStack.push_back(start);

        for (int next : graph[start]) {
            if (!visited[next]) {
                int found = findParentByDFS(next);
                parent = (visited[found] < visited[parent] ? found : parent);
            } else if (!finished[next]) {
                parent = (visited[next] < visited[parent] ? next : parent);
            }
        }

        if (parent == start) {
            sccs.emplace_back();
            int top;
            do {
                top = recStack.back();
                recStack.pop_back();
                finished[top] = true;
                sccs.back().push_back(top);
            } while (top != parent);
            std::sort(sccs.back().begin(), sccs.back().end());
        }

        return parent;
    };
    
    for (int v = 1; v <= n; ++v) {
        if (!visited[v]) {
            findParentByDFS(v);
        }
    }

    std::cout << sccs.size() << "\n";
    std::sort(sccs.begin(), sccs.end(), [](auto& s1, auto& s2) { return s1[0] < s2[0]; });
    for (auto& scc : sccs) {
        for (int v : scc) {
            std::cout << v << " ";
        }
        std::cout << "-1\n";
    }

    return 0;
}